keps/sig-storage/3333-reconcile-default-storage-class/README.md

```diff
@@ -0,0 +1,660 @@
+
+### Non-Goals
+
+* Introduce new API for default SCs.
+
+## Proposal
+
+Right now, the default SC is applied to PVCs in an admission plugin, i.e. only
+during PVC creation. Later on, PV controller will bind PVCs with
+`pvc.spec.storageClassName=nil` to PVs with `pv.spec.storageClassName=""`,
+if such a PV exists.
+
+We want to change the existing `pvc.spec.storageClassName=nil` behavior
+to always require a default SC. It would never bind to a PV with
+`pv.spec.storageClassName=""` and it would be `Pending` until a default SC
+is created. When a default SC is created, the PVC `pvc.spec.storageClassName`
+will be updated with the new default SC name in the PV controller.
+
+Any PVs with `pv.spec.storageClassName=""` will be able to bind only to a PVC
+with `pvc.spec.storageClassName=""`.
+
+This behavior should be simpler from the user perspective, without the need to
+change PV admission plugin.
+
+We plan to re-use the existing `storageclass.kubernetes.io/is-default-class`
+SC annotation.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Admin needs to change the default SC from SC1 to SC2
+1. The admin marks the current default SC1 as non-default.
+2. Another user creates PVC requesting a default SC, by leaving
+   `pvc.spec.storageClassName=nil`. The default SC does not exist at
+   this point, therefore the admission plugin leaves the PVC untouched with
+   `pvc.spec.storageClassName=nil`.
+3. The admin marks SC2 as default.
+4. PV controller, when reconciling the PVC, updates
+   `pvc.spec.storageClassName=nil` to the new SC2.
+5. PV controller uses the new SC2 when binding / provisioning the PVC.
+
+#### Story 2
+
+An installation tool wants to install Kubernetes with a CSI driver providing
+a default SC and an application that wants to use it (such as image registry).
+
+1. The installer creates PVC for the image registry first, requesting the
+   default storage class by leaving `pvc.spec.storageClassName=nil`.
+2. The installer creates a default SC.
+3. PV controller, when reconciling the PVC, updates
+   `pvc.spec.storageClassName=nil` to the new default SC.
+4. PV controller uses the new default SC when binding / provisioning the PVC.
+
+
+#### Story 3 (current behavior)
+
+User wants to provision a volume and there is one default storage class set by
+admin.
+
+1. Admin creates a default storage class.
+2. Another user creates PVC requesting a default SC, by leaving
+   `pvc.spec.storageClassName=nil`. Since the default SC already exists
+   the admission plugin changes the `nil` to a name of the default storage 
+   class.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Behavior change
+
+Currently, when a default SC is not available at PVC creation time, a PVC
+requesting a default SC (`pvc.spec.storageClassName=nil`) will keep
+`pvc.spec.storageClassName=nil` forever. Such PVC can be bound only to PV with
+`pvc.spec.storageClassName=""`.
+
+With the new behavior, the PV controller will wait until a default SC exists.
```

Comment:
Are there going to be any changes to scenarios involving pre-binding of PVs to PVCs with `claimRef` field of PV set to a PVC that will be created in the future?

Reply:
Yes, there will be changes since PVC with nil storage class are no longer bound to PVs with "" and claimRef will not have any effect on this change.

Reply:
I think we don't need to make this a behavior change - see my above comments.  If we WANT this behavior change, I need you to explain _why_ it is valuable (it is a breaking change, strictly speaking).

Reply:
I agree that retroactively applying a default seems sane and reasonable.  I just don't see why we have to include the breaking-change part.

Reply:
We could argue that even retroactively applying a default is a breaking change.

See Alternatives at the bottom of this KEP - it's possible (and actually easier) to keep PVC with `nil` binding to PV with `""`, until a default SC is created. Then all PVCs with `nil` will get the new default SC and never bind to `""`.

But it's quite confusing from user perspective. What PVC with `nil` then means? "Use default SC, if it exists at PVC creation, if it does not exist, the bind to `""`, and when there is no such PVs, then wait for them or for a new default SC."

It will be easier if `nil` always means "give me the default and wait for it if it does not exist".

Reply:
And we're willing to rewrite this KEP to keep binding `nil` <-> `""` if the current version is too breaking.

Reply:
I'm always extra paranoid about changinreally subtle semantics like this, even when they are weird.  It's just so hard to know if it will impact anyone.  Here's my thinking:

* Almost everyone uses StoirageClass
* Almost everyone uses dynamic provisioning
* Almost everyone who is NOT using dynamic provisioning is using StorageClass
* Result: the vast vast vast majority of clusters do not have and will never have PVs where `class = ""`
* Inference: anyone who DOES have such PVs probably relies on all the weird semantics of them

So leaving this functionality (PVCs with class=nil can bind to preallocated class="" PVs) is unlikely to have negative impact on people who don't want that (because almost nobody has those), but changing this is likely to have negative impact on that small subset of users who DO depend on it.

It's yucky from a code and docs POV, but that's what compat is all about.

Does my argument hold water for you?

Reply:
Thank you very much for your feedback @thockin. Those are all great points! We've changed the KEP to describe the behavior you're suggesting and moved the original proposal to the alternatives section. Please review the latest version and let us know if there's anything else we should add/change.